amr-wind/boundary_conditions/wall_models/WallFunction.cpp

```diff
@@ -11,37 +11,134 @@
 
 namespace amr_wind {
 
-WallFunction::WallFunction(const CFDSim& sim) : m_sim(sim), m_mesh(m_sim.mesh())
+WallFunction::WallFunction(CFDSim& sim)
+    : m_sim(sim), m_mesh(m_sim.mesh()), m_pa_vel(sim, m_direction)
 {
     {
         amrex::ParmParse pp("BodyForce");
         amrex::Vector<amrex::Real> body_force{{0.0, 0.0, 0.0}};
         pp.getarr("magnitude", body_force);
-        m_utau = std::sqrt(body_force[0]);
-        AMREX_ALWAYS_ASSERT_WITH_MESSAGE(
-            std::abs(body_force[1]) < 1e-16,
-            "body force in y should be zero for this wall function");
+        m_log_law.utau_mean = std::sqrt(
+            body_force[0] * body_force[0] + body_force[1] * body_force[1]);
         AMREX_ALWAYS_ASSERT_WITH_MESSAGE(
             std::abs(body_force[2]) < 1e-16,
             "body force in z should be zero for this wall function");
     }
+    {
+        amrex::ParmParse pp("transport");
+        pp.query("viscosity", m_log_law.nu);
+    }
+    const auto& geom = m_mesh.Geom(0);
+    // Reference height for log-law, this is always the first cell height
+    // NOTE: this wall function assumes that the first cell height and the
+    // last cell height are the same
+    m_log_law.zref =
+        (geom.ProbLo(m_direction) + 0.5 * geom.CellSize(m_direction));
 }
 
-VelWallFunc::VelWallFunc(Field& /*unused*/, const WallFunction& wall_func)
+VelWallFunc::VelWallFunc(Field& /*unused*/, WallFunction& wall_func)
```

Comment:
same question here about removing the const

Reply:
`wall_func` gets modified now, for the umean  update. The  new `LogLaw` object holds some state that gets updated too.